### PR TITLE
`fuel`: Defer fuel consumption for bulk operations

### LINKED ISSUE #14161: wast_tests fuzzbug: OOB bulk operations consume fuel
OSS fuzz failure: https://oss-fuzz.com/testcase-detail/4690035493109760

The following wast fails with fuel set to `18446744073709551615`:
```wast
(module
  (memory i64 1)
  (func (export "fill64") (param $dst i64) (param $len i64)
    local.get $dst
    i32.const 0
    local.get $len
    memory.fill))

;; Drains fuel to ~0 but still traps out-of-bounds correctly -> this passes.
(assert_trap (invoke "fill64" (i64.const 0) (i64.const -1)) "out of bounds")
;; No fuel left: traps "all fuel consumed" instead of "out of bounds" -> FAILS.
(assert_trap (invoke "fill64" (i64.const 0) (i64.const -1)) "out of bounds")
```

Run it with `cargo run -- wast -Wfuel=18446744073709551615 memory_fill.reduced.wast`:
```
$ cargo run -- wast -Wfuel=18446744073709551615 memory_fill.reduced.wast 
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.35s
     Running `target/debug/wasmtime wast -Wfuel=18446744073709551615 memory_fill.reduced.wast`
Error: failed to run script file 'memory_fill.reduced.wast'

Caused by:
    0: failed directive on memory_fill.reduced.wast:35
    1: expected 'out of bounds', got 'error while executing at wasm backtrace:
    0:     0x30 - wasm-function[0]
                    at ./memory_fill.reduced.wast:30:5

Caused by:
    wasm trap: all fuel consumed by WebAssembly
'
```

This is the same as the issue described by @alexcrichton here: https://github.com/bytecodealliance/wasmtime/issues/14047#issuecomment-5297814096

### PR DESCRIPTION
**Pull request by adamrk**
For bulk operations and memory/table grows, defer consuming fuel until the operation has succeed.

This fixes https://github.com/bytecodealliance/wasmtime/issues/14161.

For small statically known sizes, we immediately consume fuel, but for larger or dynamic sizes we insert a fuel/epoch check before the operation and then only consume the variable fuel amount after.

### CODE REVIEW

File: tests/all/fuel.rs [hunk 1]
```diff
@@ -185,6 +186,7 @@ fn iloop(config: &mut Config) -> Result<()> {
                     i32.const 0
                     i32.const 65536
                     memory.copy
+                    (loop)
```
Comment (adamrk) [inline tests/all/fuel.rs]: Since we're now consuming the fuel after the fuel check, a bunch of these tests need an additional fuel check inserted at the end to pass. The empty loop block does that.

File: tests/all/fuel.rs [hunk 2]
```diff
@@ -911,64 +936,6 @@ fn variable_operator_cost_follows_bounds_check(config: &mut Config) -> Result<()
     Ok(())
 }
 
-#[wasmtime_test(wasm_features(memory64), strategies(not(Winch)))]
-#[cfg_attr(miri, ignore)]
-fn memory64_variable_operator_cost_saturates(config: &mut Config) -> Result<()> {
```
Comment (adamrk) [inline tests/all/fuel.rs:916]: I'm not sure what to do with these saturating fuel tests. Now that we only consume the fuel on a successful operation we shouldn't be able to actually hit the saturation.

File: tests/all/fuel.rs (hunk 1 above)
Comment (alexcrichton) [reply tests/all/fuel.rs]: Could the fuel check move to after the operation instead of before the operation to handle this? That way while the fuel running out would be retroactively discovered it'd basically be the same thing and wouldn't need extra updates

Comment (adamrk) [reply tests/all/fuel.rs]: Yeah we can do that. I think it'll require changing some other tests which assert that we don't do a bulk operation if fuel is at 0 just before the op, but that seems fine to me.

Comment (alexcrichton) [reply tests/all/fuel.rs]: Yeah that seems ok to me, a lot of the fuel tests are about precise behavior and it's ok to tweak them if the behavior is slightly adjusted as well

File: crates/cranelift/src/func_environ.rs [hunk 3]
```diff
@@ -5021,77 +5103,123 @@ impl FuncEnvironment<'_> {
...
         const SMALL_BULK_OP_COST: i64 = 128;
         if let Some(units) = const_units
             && let Some(cost) = units.checked_mul(i64::from(cost_per_unit))
             && cost <= SMALL_BULK_OP_COST
         {
+            if should_consume_fuel {
+                self.fuel_consumed = self.fuel_consumed.saturating_add(cost);
+            }
+            return Ok(None);
+        }
```
Comment (adamrk) [inline crates/cranelift/src/func_environ.rs:5142]: One thing that's a little weird now is that we use fuel to decide if the operation is small and then that determines if we skip the epoch check after the operation. So if a user set the fuel `cost_per_unit` to 0 and ran with epochs then we wouldn't run the epoch check after a large bulk operation. Not sure if that needs to be fixed - alternatives would be to always do the epoch check, or have distinct conditions for a "small operation" when epochs are enabled.

File: crates/cranelift/src/func_environ.rs [hunk 4]
```diff
@@ -5021,77 +5103,121 @@ impl FuncEnvironment<'_> {
...
+        // Charge the deferred size-proportional fuel now that the operation is
+        // known to have succeeded.
+        let Some(DeferredBulkOp {
+            units,
+            cost_per_unit,
+        }) = deferred
+        else {
             return Ok(());
+        };
+        debug_assert!(self.tunables.consume_fuel && cost_per_unit > 0);
```
Comment (alexcrichton) [inline crates/cranelift/src/func_environ.rs]: I was thinking this debug assert was suspicious because we should able to get here with just epochs, and then I saw CI and looks like it additionally discovered that.

File: crates/cranelift/src/func_environ.rs (hunk 3 above)
Comment (alexcrichton) [reply crates/cranelift/src/func_environ.rs:5142]: That to me feels like an obscure enough edge case that it's ok. If an embedder says these operations have 0 cost then skipping the epoch check almost seems faithful to that since if it's free why bother checking epochs in more places.

File: tests/all/fuel.rs [hunk 5]
```diff
@@ -829,44 +830,53 @@ fn custom_variable_operator_cost(config: &mut Config) -> Result<()> {
     Ok(())
 }
 
-#[wasmtime_test(wasm_features(reference_types), strategies(not(Winch)))]
+#[wasmtime_test(wasm_features(bulk_memory), strategies(not(Winch)))]
 #[cfg_attr(miri, ignore)]
-fn variable_operator_cost_failed_growth(config: &mut Config) -> Result<()> {
+fn variable_operator_cost_charged_only_on_success(config: &mut Config) -> Result<()> {
```
Comment (alexcrichton) [inline tests/all/fuel.rs]: It seems useful to test a number of operators here such as the growth ones and `memory.fill`, and in retrospect I think all of this test could be added to `fuel.wast` perhaps? Trapping behavior can't be added there but it seems reasonable to me to skip testing the trapping behavior since the semantics there are sort of inherited from general translation infrastructure which may already be tested too.

File: tests/all/fuel.rs [hunk 6]
```diff
@@ -1040,15 +992,209 @@ fn fuel_around_table_grow() -> Result<()> {
...
+
+    fn fuel_consumed(self) -> u64 {
+        self.units() * (COST_MEMORY_GROW_PER_PAGE + COST_TABLE_GROW_PER_ELEMENT)
+    }
+}
+
+/// Runs a wasm which grows a memory and a table each by `size` and returns the
+/// fuel consumed. The memory and table max sizes will be set so that the grow
+/// succeeds or fails as determined by the `succeeds` parameter.
+fn grow_fuel_consumed(config: &mut Config, size: GrowSize, succeeds: bool) -> Result<u64> {
```
Comment (alexcrichton) [inline tests/all/fuel.rs]: Could the various permutations of this test all get added to `fuel.wast` instead? Naively it seems like that'd work, but I might be missing something too

File: tests/all/fuel.rs [hunk 7]
```diff
@@ -1040,15 +992,209 @@ fn fuel_around_table_grow() -> Result<()> {
...
+    assert_eq!(consumed, GROW_FIXED_FUEL);
+
+    Ok(())
+}
+
+/// Regression test for #14161. Previously this test failed because the fill
+/// operations would consume fuel.
+#[wasmtime_test(wasm_features(bulk_memory, memory64), strategies(not(Winch)))]
+#[cfg_attr(miri, ignore)]
+fn oob_memory_fill_does_not_consume_fuel(config: &mut Config) -> Result<()> {
```
Comment (alexcrichton) [inline tests/all/fuel.rs]: Is there any particular reason to single out `memory.fill` here vs other operations? If this is specifically a regression test for one of the original failures I think it's ok to skip this test since it'll be covered by fuzzing

**Review by alexcrichton — APPROVED** [review]

File: crates/cranelift/src/func_environ.rs (hunk 4 above)
Comment (adamrk) [reply crates/cranelift/src/func_environ.rs]: Yeah sorry, just fixed it.

File: tests/all/fuel.rs (hunk 7 above)
Comment (adamrk) [reply tests/all/fuel.rs]: Yeah it was just to explicitly cover the case found in fuzzing (in reduced form). I'll remove it.

File: tests/all/fuel.rs (hunk 5 above)
Comment (adamrk) [reply tests/all/fuel.rs]: Sounds good, I removed the test and added to fuel.wast so each bulk memory operation has 3 tests there for the small const, large const, and dynamic cases.

File: tests/all/fuel.rs (hunk 6 above)
Comment (adamrk) [reply tests/all/fuel.rs]: Moved these to fuel.wast as well. `table.grow` and `memory.grow` now each have 6 tests in `fuel.wast` for the 3 size cases and success/failure.

File: crates/environ/src/tunables.rs [hunk 8]
```diff
@@ -601,8 +601,7 @@ impl VariableOperatorCost {
             memory_copy_per_byte: 1,
             memory_fill_per_byte: 1,
             memory_init_per_byte: 1,
-            // `memory.grow` did not previously have a dynamic fuel charge.
-            memory_grow_per_page: 0,
+            memory_grow_per_page: 1,
```
Comment (adamrk) [inline crates/environ/src/tunables.rs]: Note: The `fuel.wast` tests use the default fuel costs which were 0 for `memory.grow`. I changed the default here because it seems like the default shouldn't be 0, but if we want to keep the default as 0 I could just change the settings `fuel.wast` uses.

**Review by alexcrichton — APPROVED** [review]

**Comment by adamrk** [comment]
Looks like the merge queue failure was a real failure from an oob test case now trapping with a different error and the test should be changed. I'm not sure why that wasn't picked up on the PR CI run?

**Comment by alexcrichton** [comment]
Oh we've got some [more information here](https://docs.wasmtime.dev/contributing-ci.html) but CI on PRs is a subset of the CI run in the merge queue (it'd be a bit too expensive to run everything on every PR)

**Comment by adamrk** [comment]
I see it was a failure on i686. Makes sense that we don't run that on every PR.

**Comment by adamrk** [comment]
So the i686 failure was actually an unrelated bug that was just uncovered by this change:
When handling `table.grow` we would trap with an OOB error if the delta couldn't fit in a native `usize`, meaning on 32-bit host a 64-bit table grow by more than `u32::MAX` would trap when it should actually return -1.

It's just a one line fix so I included it here along with an explicit test for this behavior.

**Review by alexcrichton — APPROVED** [review]
Oh nice catch!
